Implement Storer interface twice in puppy lab

lexual opened the pull request:
Fixes  #395

Review of colleague's PR #342


#### Changes proposed in this PR:

## -

lexual-anz committed:
Implement Storer interface twice in puppy lab

runnerdave commented on 06_puppy/lexual/puppy_test.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func TestStorer(t *testing.T) {
+	// ms := make(MapStore)
```
I would suggest removing these commented out code

lexual replied:
done

runnerdave commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,79 @@
+// MapStore ////
+
+type MapStore map[uint32]Puppy
+
```
you may consider documenting all exported functions

lexual replied:
done

rokane replied:
In addition to this, can you add proper documentation to all your exported types? The `MapStore` type above has a comment which is not directly on the line above, and non descriptive of what that type is. Please update this for all exported types throughout your code.

runnerdave commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,79 @@
+	return exists
+}
+
+func main() {
```
You may consider using your Storer implementations here and do a test for its output, thus you will have 100% coverage

lexual replied:
done

runnerdave reviewed (changes requested):
pretty good, only minor changes

runnerdave commented:
you may also consider improving your PR description with a list of your changes

lexual-anz committed:
Full testing coverage, better documentation

lexual-anz committed:
gofmt fixups

lexual commented:
Responded to comments now, thanks for initial review ;)

lexual commented:
@runnerdave can you mark as approved by colleague if you are happy with my changes please??

alfredxiao commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,96 @@
+
+// UpdatePuppy updates details of Puppy in store
+func (store *MapStore) UpdatePuppy(id uint32, puppy *Puppy) {
+	(*store)[id] = Puppy{puppy.ID, puppy.Breed, puppy.Colour, puppy.Value}
```
You probably need to check if id==puppy.ID first, otherwise, they could end up inconsistent. Same in syncstore

runnerdave reviewed (dismissed):
thanks for implementing the changes, I can approve now but please remember to list your changes in the description of the PR

camscale commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,96 @@
+
+// CreatePuppy stores a copy of Puppy p
+func (store *MapStore) CreatePuppy(p *Puppy) {
+	(*store)[p.ID] = Puppy{p.ID, p.Breed, p.Colour, p.Value}
```
No need to verbosely create a new puppy. Just `(*store)[p.ID] = *p` will do the job - since the map holds a value not a pointer, it will copy on the way in.

camscale commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,96 @@
+	Value  int
+}
+
+// Storer defines standard CRUD operations for Puppys
```
nitpick: doc strings are English sentences, so they should end with a full stop. See https://blog.golang.org/godoc-documenting-go-code for details. Please fix this throughout.

camscale commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,96 @@
+
+// CreatePuppy stores a copy of Puppy p
+func (store *SyncStore) CreatePuppy(p *Puppy) {
+	store.Store(p.ID, Puppy{p.ID, p.Breed, p.Colour, p.Value})
```
Same as `MapStore` - just use `*p` as the puppy value instead of verbosely creating a new one.

rokane replied:
This has not been resolved. Please address this as mentioned by @camh-anz 
```go
store.Store(p.ID, *p)
```

camscale commented on 06_puppy/lexual/puppy_test.go:
```diff
@@ -0,0 +1,72 @@
+
+func TestStorer(t *testing.T) {
+	suite.Run(t, &storerSuite{storer: &MapStore{}})
+	ss := SyncStore{}
```
I am left wondering why you run one suite on one line, but the other over two lines with a variable. Inconsistencies like this stand out and leave the reader wondering what they have missed. It is possible I'm missing something, but I can't see a reason for it.

camscale reviewed (changes requested):
Pretty close and nice looking Go code. The comment by @alfredxiao remains and should be addressed one way or another.

lexual-anz committed:
lab06: check puppy exists before updating.

Responding to @camh code review.

Check puppy exists before attempting to update it.

Added full stops to documentation sentences.

SyncStore initiliazation no longer on sepearte line in test.

rokane commented:
@lexual the changes requested from @runnerdave involving the updating of your PR description have not been addressed. Please update your description or this will not be able to be merged.

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+
+var out io.Writer = os.Stdout
+
+type Puppy struct {
```
Please include documentation for all exported types. It is good practice to get in the habit for doing this from now on.

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+
+// Storer defines standard CRUD operations for Puppys.
+type Storer interface {
+	CreatePuppy(*Puppy)
```
Might also be a good idea to return something from this `CreatPuppy` function. Otherwise how are you able to determine whether the call actually worked? Up to you, but you might want to return a bool like you did for `DeletePuppy` or maybe the created `Puppy`

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+	if !ok {
+		return nil
+	}
+	return &p
```
Little nit, and up to you, but if you flip the logic you can initialise on the if statement and save a line of code.

```go
if p, ok := (*store)[id]; ok {
  return &p
}
return nil
```

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+
+// ReadPuppy retrieves a previously stored Puppy.
+func (store *SyncStore) ReadPuppy(id uint32) *Puppy {
+	generic, ok := store.Load(id)
```
Again here another idea is to flip the logic and you can save a few lines

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+// UpdatePuppy updates details of Puppy in store.
+func (store *SyncStore) UpdatePuppy(id uint32, p *Puppy) {
+	if id == p.ID {
+		store.Store(p.ID, Puppy{p.ID, p.Breed, p.Colour, p.Value})
```
Save as what was mentioned above, you should be able to avoid creating a new `Puppy` struct and just use p

```go
store.Store(p.ID, *p)
```

rokane commented on 06_puppy/lexual/puppy_test.go:
```diff
@@ -0,0 +1,71 @@
+	r.False(didDelete)
+}
+
+func TestStorer(t *testing.T) {
```
Nice work setting up these tests.

rokane commented on 06_puppy/lexual/puppy.go:
```diff
@@ -0,0 +1,100 @@
+
+// SyncStore ////
+
+type SyncStore struct {
```
I know it wasn't directly stated in the lab description, but another handy facility the `sync` package provides is the Mutex struct.

Up to you whether you choose to do so, it might be fun, but you can integrate Mutex capabilities into your storer implementation.

Example:

```go
// SyncStore implements the Storer interface with Mutex support
type SyncStore struct {
  sync.Map
  sync.Mutex
}

// CreatePuppy creates a new Puppy and loads it into the store
func (s *SyncStore) CreatePuppy(p *Puppy) {
  s.Lock()
  defer s.Unlock()
  s.Store(p.ID, *p)  
}
```

rokane reviewed (changes requested):
Looking really good so far. I have added some comments throughout with suggestions for improvements. There are still a number of issues which have not been addressed from previous reviewers. Please also address these as well, and in future try not to resolve a conversation if the underlying issues have not been addressed.

juliaogris reviewed (commented):
Meta comment: please add a PR description and [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i HASH`.
